File: Algorithms/leetcode/Java/AlgorithmsStudy/src/main/java/com/example/ContainerWithMostWater.java
package com.example;

public class ContainerWithMostWater {
    public int maxArea(int[] height) {
        int first = 0, last = height.length - 1;
        int retval = Math.min(height[first], height[last]) * (last - first);
        while (first < last) {
            int v;
            if (height[first] <= height[last]) {
                v = height[first];
                for (++first; first < last && height[first] <= v; ++first) {
                }
            } else {
                v = height[last];
                for (--last; first < last && height[last] <= v; --last) {
                }
            }
            v = Math.min(height[first], height[last]) * (last - first);
            if (v > retval) {
                retval = v;
            }
        }
        return retval;
    }

    public static void main(String[] args) {
        int[] input = { 1, 8, 6, 2, 5, 4, 8, 3, 7 };
        var result = new ContainerWithMostWater().maxArea(input);
        System.out.println(result);
    }
}

File: Algorithms/leetcode/TypeScript/src/container.with.most.water.ts
/*
Given n non-negative integers a1, a2, ..., an , where each represents
a point at coordinate (i, ai). n vertical lines are drawn such that the
two endpoints of the line i is at (i, ai) and (i, 0). Find two lines, 
which, together with the x-axis forms a container, such that the container
contains the most water.
*/

export function maxArea(height: number[]): number {
    let retval = 0;
    let left = 0, right = height.length - 1;
    if (right > left) {
        retval = Math.min(height[left], height[right]) * (right - left);
        while (left < right) {
            if (height[left] <= height[right]) {
                let i = left + 1;
                while (i < right && height[left] >= height[i]) {
                    ++i;
                }
                let v = Math.min(height[i], height[right]) * (right - i);
                retval = Math.max(retval, v);
                left = i;
            } else {
                let i = right - 1;
                while (i > left && height[i] <= height[right]) {
                    --i;
                }
                let v = Math.min(height[left], height[i]) * (i - left);
                retval = Math.max(retval, v);
                right = i;
            }
        }
    }
    return retval;
}
